### PR TITLE
Mint User Guide

### PR DESCRIPTION
**Pull request by gzagaris**
This PR adds a user guide for Mint, consisting of:

- Sections describing core concepts and guiding design considerations. 
- A quick introductory example illustrating key capabilities.
- A tutorial with code snippets that demonstrate specific concepts and capabilities.

### CODE REVIEW

**Comment by white238** [comment]
Do we really want to be adding powerpoint slides to our repo?

**Comment by gzagaris** [comment]
@white238, since the powerpoint was used for all the figures, it makes sense to keep it with the docs. That way we can easily update the figures as the documentation evolves. 

I don't see any significant downside to keeping a few powepoint slides as part of the repo in this case.

**Comment by gzagaris** [comment]
The powerpoint file is removed.

**Comment by gzagaris** [comment]
You can now view the documentation here: https://axom.readthedocs.io/en/feature-zagaris2-mint-user-guide/axom/mint/docs/sphinx/index.html

File: src/axom/mint/docs/sphinx/sections/cell_types.rst [hunk 1]
```diff
@@ -0,0 +1,37 @@
+.. ## Copyright (c) 2017-2019, Lawrence Livermore National Security, LLC and
+.. ## other Axom Project Developers. See the top-level COPYRIGHT file for details.
+.. ##
+.. ## SPDX-License-Identifier: (BSD-3-Clause)
+
+
+.. |quad| ..image:: ../figures/quad.png
+          :scale: 100%
+          :alt: Linear Lagrangian Quadrilateral Element
```
Comment (gzagaris) [inline src/axom/mint/docs/sphinx/sections/cell_types.rst]: For some reason, I couldn't get those replacements to work properly ((used in fem.rst)  inlining those images doesn't do the right thing either. This used to work with prior to integrating the Mint user guide with unified Axom Sphinx documentation. I'll try to see if I can resolve this somehow

File: src/axom/mint/docs/sphinx/sections/getting_started.rst [hunk 2]
```diff
@@ -0,0 +1,275 @@
+.. ## Copyright (c) 2017-2019, Lawrence Livermore National Security, LLC and
+.. ## other Axom Project Developers. See the top-level COPYRIGHT file for details.
+.. ##
+.. ## SPDX-License-Identifier: (BSD-3-Clause)
+
+.. _sections/getting_started:
+
+Getting Started with Mint
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/getting_started.rst]: Thanks for this "getting started" tutorial.  This is exactly what we need.

File: src/axom/mint/docs/sphinx/sections/getting_started.rst [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+and `ParaView`_ as follows:
+
+.. literalinclude:: ../../../examples/mint_uniform_mesh.cpp
+   :start-after: sphinx_tutorial_walkthrough_vtk_start
+   :end-before: sphinx_tutorial_walkthrough_vtk_end
+   :language: C++
+   :linenos:
+
+The resulting VTK file is stored in the specified file, ``uniform_mesh.vtk``.
+A depiction of the mesh showing a plot of `Himmelblau's Function`_ computed
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/getting_started.rst]: Please move this sentence and the figure to Step 7, after the Note.  Running the example produces the `.vtk` file, which you then show off in the figure.

File: src/axom/mint/docs/sphinx/sections/mesh_representation.rst [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+   #. Link the :ref:`Topology` of the mesh to its constituent :ref:`Geometry`
+
+   #. Support one or more *degrees of freedom*, evaluated at the given node
+      location.
+
+
+Notably, the nodes of a mesh are not necessarily just the *vertices* of the mesh.
+As discussed in the :ref:`PreliminaryConcepts` section, a mesh is a
+discretization of a PDE. Recall, the primary purpose of the mesh is to define
+the dicsrete locations, in both *space* and *time*, at which the
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/mesh_representation.rst]: typo `dicsrete`

File: src/axom/mint/docs/sphinx/sections/mesh_representation.rst [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+   A mesh node, :math:`\mathcal{n_i}`, is associated with a point,
+   :math:`p_i \in \mathbb{R}^d` and provides the means to:
+
+   #. Link the :ref:`Topology` of the mesh to its constituent :ref:`Geometry`
+
+   #. Support one or more *degrees of freedom*, evaluated at the given node
+      location.
+
+
+Notably, the nodes of a mesh are not necessarily just the *vertices* of the mesh.
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/mesh_representation.rst]: Consider rephrasing this sentence to something like

Notably, the nodes of a mesh may be more than just the vertices of the mesh.

File: src/axom/mint/docs/sphinx/sections/mesh_representation.rst [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+As discussed in the :ref:`PreliminaryConcepts` section, a mesh is a
+discretization of a PDE. Recall, the primary purpose of the mesh is to define
+the dicsrete locations, in both *space* and *time*, at which the
+*unknown variables* or *degrees of freedom* of the governing PDE are evaluated.
+Depending on the numerical scheme employed and the :ref:`CellTypes` used, the
+:ref:`Nodes` of a mesh may also be located at cell, face and edge centroids.
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/mesh_representation.rst]: Consider rephrasing this sentence to something like

Depending on the numerical scheme employed and the CellTypes used, additional mesh Nodes may be located at cell, face, and edge centroids.

File: src/axom/mint/docs/sphinx/sections/mesh_representation.rst [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+.. _figs/topological_structure:
+.. figure:: ../figures/topological_structure.png
+  :align: center
+  :scale: 95%
+  :alt: Topological Structure
+
+  Hierarchical topological structure illustrating the *downward* and *upward*
+  topological connections of the constituent mesh entities supported in Mint.
+
+  The topological connections between the constituent entities of the mesh can be
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/mesh_representation.rst]: Un-indent here to fix runaway centering (since this should not be part of the figure caption)

File: src/axom/mint/docs/sphinx/sections/mesh_representation.rst [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+Cells
+""""""
+
+A cell, :math:`\mathcal{C}_i`, is given by an ordered list of :ref:`Nodes`,
+:math:`\mathcal{C}_i=\{n_0,n_1,...n_k\}`, where each entry,
+:math:`n_j \in \mathcal{C}_i`, corresponds to a
+unique node index in the mesh. The order of :ref:`Nodes` defining a cell is
+determined according to a prescribed local numbering convention for a
+particular cell type. See :numref:`figs/linearCells` and :numref:`figs/q2Cells`.
+All Mint :ref:`CellTypes` follow the `CGNS`_ standard local numbering
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/mesh_representation.rst]: This needs to link to https://cgns.github.io/CGNS_docs_current/sids/conv.html#unstructgrid.  I understand that you may wish to link to CGNS as a top-level reference, but please here link directly to the node ordering conventions.

File: src/axom/mint/docs/sphinx/sections/execution_model.rst [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+supplied to the user-specified kernel.
+
+.. note::
+
+    If not specified, the default :ref:`executionSignature` to is set to
+    ``mint::xargs::index``, which, indicates that the supplied kernel takes
+    a single argument that corresponds to the index of the corresponding
+    iteration space, i.e, the loop index.
+
+The list of available :ref:`executionSignature` options is summarized below:
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/execution_model.rst]: - Are these the only execution signatures ever needed?  
- Why did you choose this set of signatures? 
- Is there a reference or recommendation, ideally such as the link to the CGNS node orderings?
- Is it easy (or possible) to add an execution signature, should a code ever need a new one?
- What happens if you get it wrong, such as give `mint::xargs::cellids` to a Cell Traversal Function?

File: src/axom/mint/docs/sphinx/sections/fem.rst [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+integration, etc.
+
+Mint provides basic support for :ref:`sections/fem` that consists:
+
+#. :ref:`LagrangeBasis` *shape functions* for commonly employed :ref:`CellTypes`
+#. Corresponding :ref:`Quadratures` (under development)
+#. Routines for forward/inverse :ref:`IsoparametricMapping`, and
+#. Infrastructure to facilitate adding *shape functions* for
+   new :ref:`CellTypes`, as well as, to :ref:`AddABasis`.
+
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/fem.rst]: Please mention alternative libraries, or acknowledge inspiration from other software and explain (or at least mention) why `mint::FiniteElement` was necessary.

File: src/axom/mint/docs/sphinx/sections/fem.rst [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+This functionality is collectively exposed to the application through the
+``mint::FiniteElement`` class. Concrete examples illustrating the usage of the
+``mint::FiniteElement`` class within an application code are provided in the
+:ref:`femTutorial` tutorial section.
+
+A *Finite Element Basis* consists of a family of *shape functions*
+corresponding to different :ref:`CellTypes`. Mint currently supports
+Lagrange isoparametric :ref:`sections/fem`.
+
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/fem.rst]: Add an additional mention (somewhere in here) to https://cgns.github.io/CGNS_docs_current/sids/conv.html#unstructgrid, since node order is intimately connected with correct FEM computation.

File: src/axom/mint/docs/sphinx/sections/tutorial.rst [hunk 12]
```diff
@@ -0,0 +1,1255 @@
...
+capabilities in a structured and simple format.
+To quickly learn basic Mint concepts and capabilities through an illustrative
+walk-through of a complete working code example, see the
+:ref:`sections/getting_started` section. Additional code examples, based on
+Mint mini-apps, are provided in the :ref:`sections/examples` section. For
+thorough documentation of the interfaces of the various classes and functions
+in Mint, developers are advised to consult the
+`Mint Doxygen API Documentation`_, in conjunction with this
+:ref:`sections/tutorial`.
+
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/tutorial.rst]: It feels to me like there is too much repetition or duplication between (for example) mesh_types.rst and this "meshes" section of tutorial.rst.
- You can probably just remove the first paragraph from each of the sections of tutorial.rst: that would make the tutorial shorter, easier to understand, and thus more useful.
- You might also consider folding each section of the tutorial into the appropriate higher-level document so that (for example) execution_model.rst would both introduce the ideas and show examples of their use.  (I personally like the idea of folding examples in with concepts, but you're the one writing this document.)

File: src/axom/mint/docs/sphinx/sections/tutorial.rst [hunk 13]
```diff
@@ -0,0 +1,1255 @@
...
+The resulting mesh instance points to the application's buffers. Mint may be
+used to process the data e.g., :ref:`outputToVTK`
+etc. The values of the data may also be modified, however, the mesh cannot
+dynamically grow or shrink when using :ref:`ExternalStorage`.
+
+.. warning::
+
+   A mesh using :ref:`ExternalStorage` may modify the values of the application
+   data. However, the data is owned by the application that supplied the
+   external buffers. Consequently, the associated memory cannot be reallocated
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/tutorial.rst]: Please clarify to say "Mint cannot reallocate external buffers to grow or shrink the mesh."

File: src/axom/mint/examples/mint_tutorial.cpp [hunk 14]
```diff
@@ -0,0 +1,892 @@
...
+
+// sphinx_tutorial_inverse_map_start
+
+  double xr[ 2 ];
+  int status = fe.computeReferenceCoords( xc, xr );
+
+  switch( status )
+  {
+  case mint::INVERSE_MAP_FAILED:
+    std::cout << "Newton-Rapshon failed!";
```
Comment (agcapps) [inline src/axom/mint/examples/mint_tutorial.cpp]: typo

File: src/axom/mint/docs/sphinx/sections/tutorial.rst [hunk 15]
```diff
@@ -0,0 +1,1255 @@
...
+and constituent :ref:`FieldData`, which can significantly aid in debugging.
+
+.. warning::
+
+  The Legacy `VTK File Format`_ does not provide support for face-centered
+  fields. Consequently, the output consists of only the node-centered and
+  cell-centered fields of the mesh.
+
+The functionality for outputting a mesh to VTK  is provided by the
+``mint::write_vtk()`` function. This is a *free* function in the
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/tutorial.rst]: remove emphasis: change to "This is a free function in the"

**Comment by agcapps** [comment]
OK, so I'm still learning how to use the Github review-changes facility.  Most of my "comments" should have been "review comments."  Also, a few of the larger suggestions (such as re-structuring the examples) that I put in the top-level review comment should be taken as style suggestions (hopefully my wording made that clear).

File: src/axom/mint/docs/sphinx/sections/tutorial.rst (hunk 12 above)
Comment (gzagaris) [reply src/axom/mint/docs/sphinx/sections/tutorial.rst]: There is some overlap, but, that's intentional. I don't expect people to read this documentation sequentially. Some one may jump straight to the tutorial section. I wanted the tutorial to be somewhat self-contained with links back to the concepts to provide more information.

I did think of folding the tutorial with the concepts, but, I felt that it would have been more difficult to navigate and find where things are.

File: src/axom/mint/docs/sphinx/sections/execution_model.rst (hunk 9 above)
Comment (gzagaris) [reply src/axom/mint/docs/sphinx/sections/execution_model.rst]: * We can extend the execution signature to add more in the future. These are not all that we would ever need. But, it is straightforward to extend

* The execution signatures are selected based on things that I know I need now and in the near future. 

* I am not sure what you mean by the 3rd bullet. Recommendation for the execution signatures?

* Yes, it is straightforward to add a new signature. It is meant to be extensible. 

* Since these are template arguments that dispatch to runtime calls, if you pass something that is not supported you will get a compiler error. I think I've been good with using static assert to provide meaningful messages, but, maybe there is room to grow in that regard.

File: src/axom/mint/docs/sphinx/sections/fem.rst (hunk 10 above)
Comment (gzagaris) [reply src/axom/mint/docs/sphinx/sections/fem.rst]: The 1st 2 paragraphs of this section motivate this. It is beyond the scope of this user guide to explain the theory of the Finite Element method. Especially, since Mint is not a Finite Element code. But, I can add some references.

File: src/axom/mint/docs/sphinx/sections/getting_started.rst (hunk 2 above)
Comment (gzagaris) [reply src/axom/mint/docs/sphinx/sections/getting_started.rst]: Sure, glad you liked it.

File: src/axom/mint/docs/sphinx/sections/mesh_representation.rst (hunk 4 above)
Comment (gzagaris) [reply src/axom/mint/docs/sphinx/sections/mesh_representation.rst]: Good catch! Thanks.

File: src/axom/mint/docs/sphinx/sections/citations.rst [hunk 16]
```diff
@@ -0,0 +1,23 @@
+.. ## Copyright (c) 2017-2019, Lawrence Livermore National Security, LLC and
+.. ## other Axom Project Developers. See the top-level COPYRIGHT file for details.
+.. ##
+.. ## SPDX-License-Identifier: (BSD-3-Clause)
+
+
+.. _LLNL: https://www.llnl.gov/
+.. _Umpire: https://umpire.readthedocs.io/en/develop/
+.. _Mint Doxygen API Documentation: https://lc.llnl.gov/axom/docs/doxygen/axom_doxygen/html/classaxom_1_1mint_1_1Mesh.html
+.. _Axom Quick Start Guide: https://lc.llnl.gov/axom/docs/sphinx/quickstart_guide_docs/html/index.html
```
[the remark this answers is not in the document]
Comment (gzagaris) [reply src/axom/mint/docs/sphinx/sections/citations.rst]: These point to the corresponding rdt links now.

File: src/axom/mint/docs/sphinx/sections/citations.rst [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+
+
+.. _LLNL: https://www.llnl.gov/
+.. _Umpire: https://umpire.readthedocs.io/en/develop/
+.. _Mint Doxygen API Documentation: https://lc.llnl.gov/axom/docs/doxygen/axom_doxygen/html/classaxom_1_1mint_1_1Mesh.html
+.. _Axom Quick Start Guide: https://axom.readthedocs.io/en/develop/docs/sphinx/quickstart_guide/index.html
+.. _Conduit: https://llnl-conduit.readthedocs.io/en/latest/index.html
+.. _Blueprint: https://llnl-conduit.readthedocs.io/en/latest/blueprint.html
+.. _RAJA: https://raja.readthedocs.io/en/master/index.html
+.. _Axom Toolkit: https://axom.readthedocs.io/en/develop/
```
[the remark this answers is not in the document]
Comment (gzagaris) [reply src/axom/mint/docs/sphinx/sections/citations.rst]: Yes -- we need to add labels to the other sections and use that instead. Let's do that in another PR since this change should spans across components.

File: src/axom/mint/docs/sphinx/sections/execution_model.rst (hunk 9 above)
Comment (agcapps) [reply src/axom/mint/docs/sphinx/sections/execution_model.rst]: - Bullets 1, 2, 4: Thanks for the clarification.  I recommend making these points in the text, perhaps something like  "... summarized below.  We developed these signatures based on the usage patterns found in LLNL physics codes.  By design, it is straightforward to add a new execution signature when needed."
- Bullet 3: If there were a survey paper or well known reference discussing access patterns, here would be one place to put it.  As a library user, I'd like to see a reason behind the selection of stuff (here, execution signatures) offered.  If you use the sentence I suggested, it explains why Mint offers these signatures.
- Bullet 5: Please include this fact in the docs.  I'd recommend to add something like the following sentence either to the Note or the paragraph preceding it:  "If a call through the mint execution model specifies an execution signature that does not match the traversal function, a compile time error will result."

File: src/axom/mint/docs/sphinx/sections/fem.rst (hunk 10 above)
Comment (agcapps) [reply src/axom/mint/docs/sphinx/sections/fem.rst]: I feel it would help to elaborate the first two paragraphs somewhat further.  But if you're satisfied with the first two paragraphs as motivation for `mint::FiniteElement`, feel free to resolve this conversation.

Comment (gzagaris) [reply src/axom/mint/docs/sphinx/sections/fem.rst]: I am running low on bandwidth for this. I'll try to get to this in a later PR.

File: src/axom/mint/docs/sphinx/sections/execution_model.rst (hunk 9 above)
Comment (gzagaris) [reply src/axom/mint/docs/sphinx/sections/execution_model.rst]: Done.

File: src/axom/mint/docs/sphinx/sections/fem.rst (hunk 11 above)
Comment (gzagaris) [reply src/axom/mint/docs/sphinx/sections/fem.rst]: Done.

File: src/axom/mint/docs/sphinx/sections/synopsis.rst [hunk 18]
```diff
@@ -3,20 +3,13 @@
...
+parallel execution model that underpins the development of
+computational tools and numerical discretization methods. Thereby, enable
+implementations that are born *parallel* and *portable* to new and emerging
+architectures.
 
-**Salient Features**
-
-Some of the salient features of Mint include, but are not limited to, the
-following:
+**Key Features**
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/synopsis.rst]: good change

File: src/axom/mint/docs/sphinx/sections/synopsis.rst [hunk 19]
```diff
@@ -51,24 +45,28 @@ the following third-party libraries are supported:
...
-:ref:`sections/tutorial` section that are designed to demonstrate specific Mint
-concepts and capabilities in a structured and simple format. For more detailed
-documentation of the interfaces of the various classes and functions in Mint,
-developers are advised to consult the `Mint Doxygen API Documentation`_. Last,
-complete examples and code walk-throughs of mini-apps using Mint are provided in
-the :ref:`sections/examples` section.
-
-For any questions, please consult the :ref:`sections/faq` section or feel free
-to post your question to the Axom Developers mailing list at axom-dev@llnl.gov.
+This guide discusses the basic concepts and architecture of Mint.
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/synopsis.rst]: I like the separation into bullet points

File: src/axom/mint/docs/sphinx/sections/execution_model.rst [hunk 20]
```diff
@@ -9,10 +9,11 @@
...
 ----------------
 
-Mint provides a mesh-aware :ref:`sections/execution_model`, that
-is based on the `RAJA`_ programming model abstraction layer, enabling the
-implementation of computational kernels that are born *parallel* and *portable*
-to emerging architectures.
+Mint provides a mesh-aware :ref:`sections/execution_model`,
+based on the `RAJA`_ programming model abstraction layer. The
+execution model supports on-node fine-grain parallelism for mesh traversals.
+Thereby, enable the implementation of computational kernels that are born parallel
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/execution_model.rst]: I'd rephrase from "Thereby, enable the implementation ..." to "This enables the implementation ..."
